File: 2022/src/day04.rs
use anyhow::Result;
use aoc_rust_common::Solution;
use std::str::FromStr;

pub struct Day04;

#[derive(Debug, Clone, Copy)]
struct Range {
    from: i32,
    until: i32,
}

impl FromStr for Range {
    type Err = anyhow::Error;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts: Vec<i32> = s
            .split('-')
            .map(|p| {
                p.parse()
                    .map_err(|e| anyhow::anyhow!("failed to parse range part: {}", e))
            })
            .collect::<Result<Vec<i32>>>()?;
        if parts.len() < 2 {
            return Err(anyhow::anyhow!("invalid range: {}", s));
        }
        Ok(Range {
            from: parts[0],
            until: parts[1],
        })
    }
}

impl Solution for Day04 {
    fn year(&self) -> u32 {
        2022
    }
    fn day(&self) -> u32 {
        4
    }

    fn part1(&self, input: &str) -> Result<String> {
        let mut count = 0;
        for line in input.lines() {
            let (r1, r2) = parse_ranges(line)?;
            if (r1.from <= r2.from && r1.until >= r2.until)
                || (r2.from <= r1.from && r2.until >= r1.until)
            {
                count += 1;
            }
        }
        Ok((count).to_string())
    }

    fn part2(&self, input: &str) -> Result<String> {
        let mut count = 0;
        for line in input.lines() {
            let (r1, r2) = parse_ranges(line)?;
            if r1.from <= r2.until && r1.until >= r2.from {
                count += 1;
            }
        }
        Ok((count).to_string())
    }
}

fn parse_ranges(line: &str) -> Result<(Range, Range)> {
    let mut parts = line.split(',');
    let r1 = parts
        .next()
        .ok_or_else(|| anyhow::anyhow!("missing first range"))?
        .parse()?;
    let r2 = parts
        .next()
        .ok_or_else(|| anyhow::anyhow!("missing second range"))?
        .parse()?;
    Ok((r1, r2))
}

aoc_rust_common::aoc_test!(Day04);

File: golang/2024/day04.go
package main

import (
	"fmt"
	"strings"
)

type Day04 struct{}

func (d Day04) Part1(input string) string {
	grid := d.parse(input)
	count := 0
	word := "XMAS"
	rows := len(grid)
	cols := len(grid[0])

	dirs := [][2]int{
		{0, 1}, {0, -1}, {1, 0}, {-1, 0},
		{1, 1}, {1, -1}, {-1, 1}, {-1, -1},
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			for _, dir := range dirs {
				if d.check(grid, r, c, dir, word) {
					count++
				}
			}
		}
	}
	return fmt.Sprintf("%d", count)
}

func (d Day04) Part2(input string) string {
	grid := d.parse(input)
	count := 0
	rows := len(grid)
	cols := len(grid[0])

	for r := 1; r < rows-1; r++ {
		for c := 1; c < cols-1; c++ {
			if grid[r][c] == 'A' {
				// M.M
				// .A.
				// S.S
				tl := grid[r-1][c-1]
				tr := grid[r-1][c+1]
				bl := grid[r+1][c-1]
				br := grid[r+1][c+1]

				if ((tl == 'M' && br == 'S') || (tl == 'S' && br == 'M')) &&
					((tr == 'M' && bl == 'S') || (tr == 'S' && bl == 'M')) {
					count++
				}
			}
		}
	}
	return fmt.Sprintf("%d", count)
}

func (d Day04) parse(input string) [][]rune {
	var grid [][]rune
	for _, line := range strings.Split(input, "\n") {
		if line == "" {
			continue
		}
		grid = append(grid, []rune(line))
	}
	return grid
}

func (d Day04) check(grid [][]rune, r, c int, dir [2]int, word string) bool {
	for i := 0; i < len(word); i++ {
		nr, nc := r+dir[0]*i, c+dir[1]*i
		if nr < 0 || nr >= len(grid) || nc < 0 || nc >= len(grid[0]) || grid[nr][nc] != rune(word[i]) {
			return false
		}
	}
	return true
}
